### ISSUE TITLE
Model Evaluation

### ISSUE THREAD

**Issue by miriamkw** [miriamkw/GluPredKit#58]
GluPredKit refactor for enhanced and strict model evaluation:

Refactor:
- [x] All models should be multiple output
    - [x] Create multioutput models
        - [x] Scikit models
        - [x] Neural networks
    - [x] Create multioutput preprocessor
    - [x] Update all CLI commands to support multiple outputs
    - [x] Delete outdated files
- [x] What-if events should be default
- [x] Fix plug-and-play error: ohio should also be able to be used with normal preprocessor
- [x] Update README to the refactoring

Configuration:
- [ ] Number of time lagged events should be specified for each feature (list instead of integer)
- [x] Configuration has one prediction horizon
- [x] Test size should be removed to parsers, CLI command instead. In that way, we can make preprocessors properly plug-and-play
- [x] Maybe the type of numerical transformation should be an input to the configuration instead of creating a new preprocessor for every time you want to use something different

Generate pdf reports for model evaluation:
- [x] Create CLI command to generate a report
- [x] Add a text at top or bottom that refers to GluPredKit ("This report is generated using GluPredKit...")
- [x] Add dataset information
- [x] Add configuration information
- [x] Add traditional evaluation metrics
- [x] Add run time evaluation 
- [x] Insulin and carbohdyrate effect evaluation: Use PCP or the other number (S..). Idea for Partial Dependence Plot: x-axis is prediction horizon, y axis is predicted blood glucose, but then we can have many lines with labels like (UI = 1, UI=2...).
- [x] Generate a report that compare models - have in mind that datasets usually will contain many subjects.

Models:
- [x] Delete models that don't fit our requirements? Or define included models based on inclusion criterion. Maybe models from the literature that have included their source code for example? Use Doyoungs work as help?
- [x] Uva padova model
- [x] T1D loop model
- [x] Tidepool Loop model

Parsers:
- [x] Parsers should process data with the new set of features that we agree upon with Sam
- [ ] Add OpenAPS as well
- [x] Should assume that all the dataset subjects are in the dataset initially
- [x] Parsers should already have separated train and test data to avoid data leakage
    - [x] Remove test data fraction from configuration
    - [x] Add it to parsing CLI command instead?
- [x] Make sure that all parsers use the same unit for basal rates

Metrics:
- [x] Add MARD number

Testing:
- [x] Make sure that all components have tests and that running the tests is working

General improvements:
- [x] Be clearer about the expected input and output formats for each module
- [x] Add citation information in README

Checklist before merge
- [ ] Clean up the code properly
- [ ] Make sure that all the CLI commands are working on both Win and Mac
- [x] Run tests
- [x] Update README to reflect all of the changes

**Comment by miriamkw** [comment]
Subtasks adding Loop model:
- [ ] Use the 1800 rule ++ to fit the therapy settings to individuals
- [ ] Figure out how to handle smoothly the history lengths when calculating metrics (maybe loop prediction algorithm also should use the "time lagged features" from configuration)
- [ ] Add future carbs and insulin
- [ ] Make sure that parsers add basals correctly and loop model uses basals correctly wrt units

**Comment by miriamkw** [comment]
Subtasks adding whole Ohio dataset evaluation:
- [ ] Parsing: Storing the dataset in one file, 
- [ ] CLI: Parse the whole dataset, store in one file
- [ ] Model: Train one model each individual? Use the individual models for model evaluation?
- [ ] Generated report

**Comment by miriamkw** [comment]
Subtask adding the single model evaluation:
- [ ] Add a test-model function to store predictions and configurations so that you can prototype the generated reports more quickly
- [ ] store predictions for a ridge and a loop model
- [ ] train a loop model with optimizing parameters for a loop model to make the predictions and distributions more realistic
- [ ] Add a zero-hold algorithm and store predictions
- [ ] Add a naive linear regressor and store predictions
- [ ] Add a 3rd or 4th order arx and store predictions
- [ ] Model accuracy: 
    - [ ] In the plot, compare to the naive models
    - [ ] Add metrics: MARE, ME/MRE
    - [ ] Plot parkes error grid (pick max 6 within the prediction horizon)
    - [ ] Write parkes error grid results in a table
- [ ] Glycemia Detection:
    - [ ] Hypo: Plot accuracy, MCC across the prediction horizon
    - [ ] Hyper: Plot accuracy, MCC across the prediction horizon
    - [ ] Total: Make a table, calculate overall results
- [ ] Predicted distribution: Remove the second plot, add indication from the test data, maybe add an interpretation of it in text format? Make it visually clear what is the preferred range.
- [ ] Physiological Alignment: 
    - [ ] Calculate SHAP numbers. They provide positive or negative impact indication, and magnitude.
    - [ ] Plot a matrix of SHAP numbers across the prediction horizon and for different time lags
    - [ ] Add color coding to the matrix so we can clearly see the impact. 
    - [ ] Make the color red if it contradicts the physiological reality
    - [ ] Make an interpretation of the matrix in writing? Or add the plot for a physiological model.
- [ ] Finish the first page overview of the model evaluation
- [ ] Write a document explaining all of the plots and the numbers and the benchmark naive algorithms
- [ ] Clean up the code so that the cli file is very understandable, and making a helper file that can be more messy

**Comment by miriamkw** [comment]
Subtask adding the single model evaluation:
- [ ] Error grid analysis cost functions

**Comment by miriamkw** [comment]
JOSS To dos:
- [ ] Parsers
    - [ ] Separate into test and train data immediately
    - [ ] Remove it from configs
    - [ ] Make sure that basal is always in the same format
- [ ] Models
    - [ ] Make a set of models supporting multiple outputs
    - [ ] Fix the loop model accuracy (?)
    - [ ] remove tensorflow models? 
- [ ] Feature scaling technique should be defined in the model configuration
- [ ] CLI
    - [ ] Make sure the CLI is updated to the new refactor
    - [ ] Update README
- [ ] README
    - [ ] Define input and output formats for each module ("Community Contributions")
- [ ] Dependencies: Make sure they are not too strict, and that all dependencies are there
- [ ] Code cleanup: Remove all unneccessary.prints and comments in the code.
- [ ] Error testing with both windows and mac, and both with pypi and the repo directly
- [ ] Add tests
- [ ] Add CI (Sam)
- [ ] Merge to main
- [ ] Push to PyPi
- [ ] Check the JOSS checklist

**Comment by miriamkw** [comment]
JOSS specifics, last to dos:
- [ ] Write good documentation and tests
- [ ] Create a branch from the main branch, and add the paper.md with bib there
- [ ] Send the paper for review from coauthors
- [ ] Check that the paper compiles using github action
- [ ] README should contain:
    - [ ] A statement of need: which problems the software is designed to solve, who the target audience is, and its relation to other work
    - [ ] installation instructions (pip install ...)
    - [ ] examples of how to use the software
    - [ ] API documentation: all functions are documented including example inputs and outputs
    - [ ] Community guidelines: clear guidelines for third-parties wishing to contribute, report issues or problems, or seek support
- [ ] Tests: an automated test suite hooked up to continuous integration (github actions, circle CI or similar)
- [ ] Sample dataset: There should be a sample dataset because reviewers are expected to install the software and verify core functionality

**Comment by miriamkw** [comment]
Tests:
- [ ] Read into unit tests vs integration tests
- [ ] Finish testing the cli commands, make sure that all the models are running and so on
- [ ] Make unit tests on specific functions
    - [ ] Determine which functions are called most often by using a profiler
    - [ ] Test those functions
